[deutschebahn] Initial contribution: New binding for DeutscheBahn Fahrplan (#11384)
openhab/openhab-addons: bundles/org.openhab.binding.deutschebahn/README.md

# DeutscheBahn Binding

The DeutscheBahn Binding provides the latest timetable information for all trains that arrive or depart at a specific train station, including live information for delays and changes in timetable.

> The informations are requested from the timetable api of DeutscheBahn developer portal, so you'll need a (free) developer account to use this binding.

Reviewer: ```suggestion
The information are requested from the timetable api of DeutscheBahn developer portal, so you'll need a (free) developer account to use this binding.
```

Author: fixed

## Supported Things

- **timetable** The timetable bridge connects to the timetable api and provides information for the next trains that will arrive or depart at the configured station.
- **train** The train thing represents one trains within the configured timetable. This may be an arrival or a departure.

## Thing Configuration

### Generate Access-Key for timetable API

To configure a timetable you first need to register at DeutscheBahn developer portal and register for timetable API to get an access key.

1. Go to [DeutscheBahn Developer](https://developer.deutschebahn.com)
2. Register new account or login with an existing one
3. If no application is configured yet (check Tab "Meine Anwendungen") create a new application. Only the name is required, any other fields can be left blank.
4. Go to APIs - Timetables v1 (may be displayed on second page)
5. Choose your previously created application and hit "Abonnieren"
6. In confirmation-dialog choose "Wechsel zu meine Abonnements"
7. Create an access key for the production environment by hitting "Schlüssel Erstellen"

> 8. Copy the "Zugangstoken". This is required to access the api from openhab.

Reviewer: ```suggestion
8. Copy the "Zugangstoken". This is required to access the api from openHAB.
```

Author: fixed

### Determine the EVA-No of your station

> For the selection of the station within openhab you need the eva nr. of the station.

Reviewer: ```suggestion
For the selection of the station within openHAB you need the eva no. of the station.
```

Author: fixed

You can look up the number within the csv file available at [Haltestellendaten](https://data.deutschebahn.com/dataset.tags.EVA-Nr..html).

### Configure timetable bridge

> With access key for developer portal and eva nr. of your station you'r ready to configure a timetable (bridge) for this station.

Reviewer: ```suggestion
With access key for developer portal and eva no. of your station you're ready to configure a timetable (bridge) for this station.
```

Author: fixed

In addition you can configure if only arrivals, only departures or all trains should be contained within the timetable.

**timetable** parameters:

| Property | Default | Required | Description |
|-|-|-|-|
| `accessToken` | | Yes | The access token for the timetable api within the developer portal of DeutscheBahn. |
| `evaNo` | | Yes | The eva nr. of the train station for which the timetable will be requested.|
| `trainFilter` |  | Yes | Selects the trains that will be displayed in the timetable. Either only arrivals, only departures or all trains can be displayed. |

### Configuring the trains

Once you've created the timetable you can add train-things that represent the trains within this timetable.
Each train represents one position within the timetable. For example: If you configure a train with position 1 this will be
the next train that arrives / departs at the given station. Position 2 will be the second one, and so on. If you want to
show the next 4 trains for a station, create 4 things with positions 1 to 4.

> **Attention:** The timetable api only provides data for the next 18 hours. If the timetable contains less trains entries than you've created

Reviewer: ```suggestion
**Attention:** The timetable api only provides data for the next 18 hours. If the timetable contains less train entries than you've created
```

Author: fixed

> train things the channels of these trains will be undefined.

Reviewer: ```suggestion
train things, the channels of these trains will be undefined.
```

**train** parameters:

| Property | Default | Required | Description |
|-|-|-|-|
| `position` | | Yes | The position of the train within the timetable. |

## Channels

Each train has a set of channels, that provides access to any information served by the timetable API. A detailed description of the values and their meaning can be found within
the [Timetables V1 API Description](https://developer.deutschebahn.com/store/apis/info?name=Timetables&version=v1&provider=DBOpenData&).

> The informations are grouped into three channel-groups:

Reviewer: ```suggestion
The information are grouped into three channel-groups:
```

> The first channel group (trip) contains all informations for the trip of the train, for example the category (like ICE, RE, S).

Reviewer: ```suggestion
The first channel group (trip) contains all information for the trip of the train, for example the category (like ICE, RE, S).
```

> The second and third channel group contains informations about the the arrival and the departure of the train at the given station. 

Reviewer: ```suggestion
The second and third channel group contains information about the the arrival and the departure of the train at the given station. 
```

Both of the groups may provide an 'UNDEF' channel value, when the train does not arrive / depart at this station
(due it starts or ends at the given station). If you have configured your timetable to contain only departures (with property trainFilter) the departure channel values will always be defined 
and if you have selected only arrivals the arrival channel values will always be defined.

> Channels will have an 'NULL' channel value, when corresponding attribute is not set.

Reviewer: ```suggestion
Channels will have a 'NULL' channel value, when the corresponding attribute is not set.
```

> Basically most informations are available as planned and changed value. This allows to easy display changed values (for example the delay or changed platform).

Reviewer: ```suggestion
Basically most information are available as planned and changed value. This allows to easy display changed values (for example the delay or changed platform).
```

Author: fixed

| filter-flags | String | Provides the filter flags. |
| trip-type | String | Provides the type of the trip. |
| owner | String | Provides the owner of the train. A unique short-form and only intended to map a trip to specific evu (EisenbahnVerkehrsUnternehmen). |

**Channels for arrival / departure**
| channel  | type   | description                  |
|----------|--------|------------------------------|
| planned-path  | String | Provides the planned path of a train. |
| changed-path  | String | Provides the changed path of a train. |
| planned-platform  | String | Provides the planned platform of a train.  |
| changed-platform | String | Provides the changed platform of a train. |
| planned-time | DateTime | Provides the planned time of a train. |
| changed-time | DateTime | Provides the changed time of a train. |
| planned-status | String | Provides the planned status (planned, added, cancelled) of a train. |
| changed-status | String | Provides the changed status (planned, added, cancelled) of a train. |
| cancellation-time | DateTime | Time when the cancellation of this stop was created. |
| line | String | The line of the train.  |

> | messages | String | Messages for this train. Contains all translated codes from the messages of the selected train stop. Multiple messages will be separated with an single dash. |

Reviewer: ```suggestion
| messages | String | Messages for this train. Contains all translated codes from the messages of the selected train stop. Multiple messages will be separated with a single dash. |
```

Author: fixed

> | hidden | Switch | On if the event should not be shown on WBT because travellers are not supposed to enter or exit the train at this stop. |

Reviewer: WBT?

Author: The descriptions are taken from the original api documentation, so even i don't know what WBT means. I think it might be "Wall Board Timetable" or something else. So due this has no relevance here, i removed it.

> | wings | String | A sequence of trip id separated by the pipe symbols. |

Reviewer: ```suggestion
| wings | String | A sequence of trip id separated by pipe symbols. |
```

Author: fixed

          - component: f7-col
            config:
              width: 50
            slots:
              default:
                - component: Label
                  config:
                    text: "=items[props.changed_intermediate_stations].state != 'NULL' ? items[props.changed_intermediate_stations].state : items[props.planned_intermediate_stations].state"
                    style:
                      color: "=items[props.changed_intermediate_stations].state != 'NULL' ? 'red' : ''"
          - component: f7-col
            config:
              width: 35
            slots:
              default:
                - component: Label
                  config:
                    text: "=items[props.changed_final_station].state != 'NULL' ? items[props.changed_final_station].state : items[props.planned_final_station].state"
                    style:
                      color: "=items[props.changed_final_station].state != 'NULL' ? 'red' : ''"
                      font-weight: bold
                      text-align: right
```

Using the widget for displaying the next four departures:

> ![Departures Hannover HBF](doc/images/Abfahrten_HannoverHBF.png "openHAB page with four widgets displaying the next departures at Hannover HBF")

Reviewer: I think images need to reside directly under `doc/` to be rendered correctly on the OH webpage.

Author: fixed
